### PR TITLE
luci-app-olsr: migrate to JavaScript-based implementation

### LINKED ISSUE #5363: luci-app-olsr: port to js
This is a request to have luci-app-olsr ported to javascript.  The status pages, in particular, would benefit greatly from client-side rendering.

### PR DESCRIPTION
**Pull request by Ayushmanwebdeveloper**
https://github.com/openwrt/luci/assets/91346165/ec75e0a1-942f-4197-830e-130957135dc6

Hi! I've almost completed the migration of luci-app-olsr to JavaScript. Most of the features are functioning just like they did in the old Lua app. There are only two minor bugs remaining in the status views, which I will make sure to fix soon.

You can review the changes and please provide any suggestions or improvements. Since this is a very large application, I believe there is ample scope for feedback.

During the migration, I implemented a new RPCD and made changes to how plugins and their configurations were handled in the old app. I have implemented plugin configuration using the previous approach used for interfaces.

Most of the work is complete and functional. I am also uploading a screencast for reference. Fixes #5363 

Thank you for your time!

### CODE REVIEW

**Comment by Ayushmanwebdeveloper** [comment]
And one issue is the null that you can see in the screencast in some of the status views, which is actually the hostname that I'm attempting to obtain using hosthints.getHostnameByIPAddr(v.gateway). We're getting the data and hosthints successfully, as can be seen in the screenshot, but it's still returning null, meaning that the hostname not known [](https://openwrt.github.io/luci/jsapi/LuCI.network.Hosts.html#getHostnameByIPAddr).
 It's used in other apps like this. But I'll attempt to fix it as soon as possible.

image

```
					network
						.getHostHints()
						.then(function (hosthints) {
							modifiedData = data.map(function (v) {
								if (resolveVal === '1') {
									var hostname = hosthints.getHostnameByIPAddr(v.gateway);
```

**Comment by jow-** [comment]
Looks good! You should deal with the empty hostnames by doing something like `hostname || '?'` when outputting them.

I see you do something like `x = '<a href="http://' + v.hostname + '/blah">' + v.hostname + '</a>';` in various places. This will allow for XSS injections if someone broadcasts packets containing forged hostnames into the network. Ideally use LuCI's builtin format function together with the `%h` and `%q` placeholder which take care of HTML- and quote-escaping their arguments respectively:

 `x = '<a href="http://%q/blah">%h</a>'.format(v.hostname, v.hostname);`

**Comment by Ayushmanwebdeveloper** [comment]
> Looks good! You should deal with the empty hostnames by doing something like `hostname || '?'` when outputting them.
> 
> I see you do something like `x = '<a href="http://' + v.hostname + '/blah">' + v.hostname + '</a>';` in various places. This will allow for XSS injections if someone broadcasts packets containing forged hostnames into the network. Ideally use LuCI's builtin format function together with the `%h` and `%q` placeholder which take care of HTML- and quote-escaping their arguments respectively:
> 
> `x = '<a href="http://%q/blah">%h</a>'.format(v.hostname, v.hostname);`

I fixed that for every instance of the hostname.

**Comment by Ayushmanwebdeveloper** [comment]
image
image


Hi! @jow- @aparcar @feckert  Now it's ready for review; I fixed all of the bugs, including the **Hostname**. Everything is functional, and we tested the package for the last 2 weeks and tried to resolve every issue. Could you please review? Thanks

File: applications/luci-app-olsr/Makefile [hunk 1]
```diff
@@ -7,7 +7,7 @@
 include $(TOPDIR)/rules.mk
 
 LUCI_TITLE:=OLSR configuration and status module
-LUCI_DEPENDS:=+luci-compat +olsrd +olsrd-mod-jsoninfo +luci-lib-json
+LUCI_DEPENDS:=+luci-compat +olsrd
```
[the remark this answers is not in the document]
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr/Makefile:10]: Okay! I will remove luci-compat 👍

File: applications/luci-app-olsr/root/usr/share/rpcd/acl.d/luci-app-olsr-unauthenticated.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+				"network": ["get_proto_handlers"],
+				"olsrd": ["olsrd_jsoninfo"],
+				"olsrd6": ["olsrd_jsoninfo"],
+				"olsrinfo": ["getjsondata", "hasipip"],
+				"file": ["read"],
+				"iwinfo": ["assoclist"]
+
+			},
+			"file": {
+				"/tmp/hosts/olsr.ipv4": [ "read" ]
```
[the remark this answers is not in the document]
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr/root/usr/share/rpcd/acl.d/luci-app-olsr-unauthenticated.json:18]: Okay! Sure, we can do that. I'll try to fix that soon.

File: applications/luci-app-olsr/Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr/Makefile]: I made all of the mentioned changes.

File: applications/luci-app-olsr/root/usr/share/rpcd/acl.d/luci-app-olsr-unauthenticated.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr/root/usr/share/rpcd/acl.d/luci-app-olsr-unauthenticated.json]: I implemented it as rpcd and also tested it.

**Comment by andibraeu** [comment]
@jow- @aparcar can you please take a look on that PR?

for me it looks good

**Comment by jow-** [comment]
@andibraeu - I don't have the means to runtime test but if you say it's fine then I'll go ahead and merge it. @Ayushmanwebdeveloper - thanks a lot for your effort!

**Comment by thuehn** [comment]
Hi all,

Just flashing our Freifunk Evernet last nigth with current OpenWrt trunk and Luci we run into the following issue, that might belong to this commit:

after flashing trunk, the LUCI webinterface did only show olsr status - no menu for admin/network & co anymore:

image


After deinstalling luci-app-olsr the Luci webinterface went back to a normal menu:

image

Any hints ?

Greets Bluse

**Comment by Ayushmanwebdeveloper** [comment]
> Hi all,
> 
> Just flashing our Freifunk Evernet last nigth with current OpenWrt trunk and Luci we run into the following issue, that might belong to this commit:
> 
> after flashing trunk, the LUCI webinterface did only show olsr status - no menu for admin/network & co anymore:
> 
> image
> After deinstalling luci-app-olsr the Luci webinterface went back to a normal menu:
> 
> image
> @andibraeu @jow- any hints from the experts ? (seems like something is overloaded by luci-app-olsr)
> 
> Greets Bluse

Hello! We migrated the app such that status views are exposed to the public, which means they can be accessed without authentication.

Particularly in this openwrt theme, you can return to admin by choosing **Administration** in the bottom right area of the page.

image

In the Freifunk or Material theme everything works okay as shown in my screencast.

**Comment by thuehn** [comment]
Hiho

The button OLSR | ADMINISTRATION is nether on theme material nor on bootstrap visible on our side:

image

image

There is no exit to the admin web interface available as long as luci-app-olsr is installed.. ?

Greetz Bluse

**Pull request merged**